File: Data Structures & Algorithms/add-two-numbers/submission-2.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} l1
     * @param {ListNode} l2
     * @return {ListNode}
     */
    addTwoNumbers(l1, l2) {
        let resHead = new ListNode();
        let curr = resHead;
        let carry = 0;
        while(l1 && l2) {
            let sum = l1.val + l2.val + carry;
            if (sum < 10) {
                curr.val = sum;
                carry = 0;
            } else {
                carry = 1;
                curr.val = sum % 10;
            }
            if (l1.next || l2.next) {
                curr.next = new ListNode();
                curr = curr.next;
            }
            l1 = l1.next;
            l2 = l2.next;
        }
        while (l1) {
            curr = l1.val;
            if (l1.next) {
                curr.next = new ListNode();
                curr = curr.next;
            }
            l1 = l1.next;
        }
        while (l2) {
            curr = l2.val;
            if (l2.next) {
                curr.next = new ListNode();
                curr = curr.next;
            }
            l2 = l2.next;
        }
        return resHead;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class NodeWrapper:
    def __init__(self, node):
        self.node = node
    
    def __lt__(self,other):
        return self.node.val < other.node.val

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []
        for ll in lists:
            if ll:
                heapq.heappush(heap, NodeWrapper(ll))
        
        dummyHead = ListNode(0, None)
        curr = dummyHead
        while heap:
            nodeWrapper = heapq.heappop(heap)
            curr.next = nodeWrapper.node
            curr = curr.next
            if nodeWrapper.node.next:
                heapq.heappush(heap, NodeWrapper(nodeWrapper.node.next))
        return dummyHead.next